text/00xx-smart-contracts-coretime-chain.md

```diff
@@ -0,0 +1,102 @@
+
+## Stakeholders
+
+Primary stakeholders include:
+- Developers working on the Coretime chain.
+- Users who want to deploy decentralized applications on the Coretime chain.
+- Community members interested in expanding the capabilities of the Coretime chain.
+- Secondary Coretime marketplaces.
+
+## Explanation
+
+This RFC introduces the following key components:
+
+1. **Smart Contract Support**:
+    - Implement support for deploying and executing smart contracts on the Coretime chain.
+    - Use a well-established smart contract platform, such as Ethereum’s Solidity or Polkadot's Ink!, to ensure compatibility and ease of use.
+
+2. **Storage and Execution**:
+    - Define a storage structure for smart contracts and their associated data.
+    - Ensure efficient and secure execution of smart contracts, with proper resource management and gas fee mechanisms.
+
+3. **Integration with Existing Pallets**:
+    - Ensure that smart contracts can interact with existing pallets on the Coretime chain, such as the broker pallet.
+    - Provide APIs and interfaces for seamless integration and interaction.
+
+4. **Security and Auditing**:
+    - Implement robust security measures to prevent vulnerabilities and exploits in smart contracts.
+    - Conduct thorough security audits and testing before deployment.
+
+## Drawbacks
+
+There are several drawbacks to consider:
+- **Complexity**: Adding smart contracts introduces significant complexity to the Coretime chain, which may increase maintenance overhead and the potential for bugs.
+- **Performance**: The execution of smart contracts can be resource-intensive, potentially affecting the performance of the Coretime chain.
+- **Security**: Smart contracts are prone to vulnerabilities and exploits, necessitating rigorous security measures and continuous monitoring.
+
+## Testing, Security, and Privacy
+
+### Testing
+- Comprehensive unit tests and integration tests should be developed to ensure the correct functionality of smart contracts.
+- Test scenarios should include various use cases and edge cases to validate the robustness of the implementation.
+
+### Security
+- Security audits should be performed to identify and mitigate vulnerabilities.
+- Implement best practices for smart contract development to minimize the risk of exploits.
+- Continuous monitoring and updates will be necessary to address new security threats.
+
+### Privacy
+- The proposal does not introduce new privacy concerns as it extends existing functionalities with programmable logic.
+
+## Performance, Ergonomics, and Compatibility
+
+### Performance
+- The introduction of smart contracts may impact performance due to the additional computational overhead.
+- Optimization techniques, such as efficient gas fee mechanisms and resource management, should be employed to minimize performance degradation.
```

Comment:
This should be pretty easy to mitigate by giving contracts a custom `DispatchClass` and limiting how much of the block they can consume, so we ensure some amount is always reserved for "core" Coretime functionality.